### PR TITLE
Update react version

### PR DESCRIPTION
**Pull request by tiagobarros01**
## Summary

* Updates React version to 18;
* Updates to new React Render format;
* Updates types;
* Changes tests to use `renderHook` from `@testing-library/react`

### Notes

* The `useCroct` test, continues using the `renderHook` from `@testing-library/react-hooks` because the return from the `renderHook` from `@testing-library/react` is different and did not execute the `expect`

### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/CroctProvider.tsx [hunk 1]
```diff
@@ -2,7 +2,9 @@ import {createContext, FunctionComponent, ReactElement, useContext, useEffect, u
 import {Configuration, Plug} from '@croct/plug';
 import {croct} from './ssr-polyfills';
 
-export type CroctProviderProps = Configuration & Required<Pick<Configuration, 'appId'>>;
+export type CroctProviderProps = Configuration & Required<Pick<Configuration, 'appId'>> & {
+    children?: ReactElement,
```
Comment (marcospassos) [inline src/CroctProvider.tsx:6]:
```suggestion
    children?: ReactNode,
```

File: src/hooks/useCroct.ssr.test.tsx [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-import {ReactNode} from 'react';
+import {ReactElement} from 'react';
 import {useCroct} from './useCroct';
 import {CroctProvider} from '../CroctProvider';
 
 describe('useCroct', () => {
     it('should not fail on server-side rendering', () => {
         const {result} = renderHook(() => useCroct(), {
-            wrapper: ({children}: {children?: ReactNode}) => (
+            wrapper: ({children}: {children?: ReactElement}) => (
```
Comment (marcospassos) [inline src/hooks/useCroct.ssr.test.tsx:9]: is `{children?: ReactElement}` still needed?

File: src/hooks/useCroct.test.tsx [hunk 3]
```diff
@@ -17,7 +18,9 @@ describe('useCroct', () => {
 
     it('should return the Plug instance', () => {
         const {result} = renderHook(() => useCroct(), {
-            wrapper: ({children}) => (<CroctContext.Provider value={{plug: croct}}>{children}</CroctContext.Provider>),
+            wrapper: ({children}: {children: ReactElement}) => (
```
Comment (marcospassos) [inline src/hooks/useCroct.test.tsx:21]: same

**Review by marcospassos — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -54,7 +54,7 @@
     "@testing-library/react-hooks": "^8.0.0",
     "@types/jest": "^27.5.1",
     "@types/node": "^16.0.0",
-    "@types/react": "^17.0.0",
```
Comment (marcospassos) [inline package.json:57]: shouldn't we remove `@testing-library/react-hooks`?

**Review by marcospassos — CHANGES_REQUESTED** [review]

File: src/CroctProvider.tsx (hunk 1 above)
Comment (joaopaulomoraes) [reply src/CroctProvider.tsx:6]: Or you can use the generic `PropsWithChildren` when using a `FunctionComponent`.

```ts
// react/index.d.ts
type PropsWithChildren<P> = P & { children?: ReactNode };
```

eg: 

```tsx
export const  CroctProvider: FunctionComponent<PropsWithChildren<CroctProviderProps>> = (...) => {};
```

> I agree with `ReactNode` and maybe you don't need to use `FunctionComponent`.

File: examples/csr-app/src/index.tsx [hunk 5]
```diff
@@ -1,16 +1,15 @@
 import React from 'react';
-import ReactDOM from 'react-dom';
+import ReactDOM from 'react-dom/client';
 import {CroctProvider} from '@croct/plug-react';
 import App from './App';
 import reportWebVitals from './reportWebVitals';
 
-ReactDOM.render(
+ReactDOM.createRoot(document.getElementById('root')!).render(
```
Comment (joaopaulomoraes) [inline examples/csr-app/src/index.tsx:7]: 💡 You can use `document.getElementById('root') as HTMLElement`

File: src/CroctProvider.tsx (hunk 1 above)
Comment (marcospassos) [reply src/CroctProvider.tsx:6]: or just:

``ts
type CroctProviderProps = PropsWithChildren<...>;
``

File: src/hooks/useCroct.ssr.test.tsx [hunk 6]
```diff
@@ -1,12 +1,12 @@
 import {renderHook} from '@testing-library/react-hooks/server';
```
[the remark this answers is not in the document]
Comment (tiagobarros01) [reply src/hooks/useCroct.ssr.test.tsx:1]: As I explained in the PR note, in some tests like the `server-side-rendering` it wasn't possible to use the renderHook from @testing-library/react, so we're using both libs.
But the correct thing is to expect @testing-library/react-hooks to support version 18 of React, as it's in the library documentation, and all tests will use it.

[react-hooks-testing-library](https://github.com/testing-library/react-hooks-testing-library)

File: package.json (hunk 4 above)
Comment (tiagobarros01) [reply package.json:57]: [Same](https://github.com/croct-tech/plug-react/pull/178#discussion_r899998314)

File: src/hooks/useCroct.ssr.test.tsx (hunk 2 above)
Comment (tiagobarros01) [reply src/hooks/useCroct.ssr.test.tsx]: Yes, because `children` cannot be `undefined`

File: src/hooks/useCroct.test.tsx (hunk 3 above)
Comment (tiagobarros01) [reply src/hooks/useCroct.test.tsx]: [Same](https://github.com/croct-tech/plug-react/pull/178#discussion_r900003192)

File: src/hooks/useCroct.ssr.test.tsx [hunk 7]
```diff
@@ -1,12 +1,17 @@
...
 import {ReactNode} from 'react';
+import {Plug} from '@croct/plug';
 import {useCroct} from './useCroct';
 import {CroctProvider} from '../CroctProvider';
 
+type RenderHookProps = {
```
Comment (marcospassos) [inline src/hooks/useCroct.ssr.test.tsx:7]: Is it really needed? What is the error without this?

Comment (tiagobarros01) [reply src/hooks/useCroct.ssr.test.tsx:7]: `Type '({ children }: { children: any; }) => Element' is not assignable to type 'WrapperComponent<unknown> | undefined'.`

File: src/hooks/useCroct.ssr.test.tsx [hunk 8]
```diff
@@ -1,12 +1,17 @@
...
+    children: ReactNode,
+};
+
 describe('useCroct', () => {
     it('should not fail on server-side rendering', () => {
-        const {result} = renderHook(() => useCroct(), {
-            wrapper: ({children}: {children?: ReactNode}) => (
+        const {result} = renderHook<RenderHookProps, Plug>(() => useCroct(), {
```
Comment (georgekaran) [inline src/hooks/useCroct.ssr.test.tsx:13]: Shouldn't this be `CroctProviderProps`?

```suggestion
        const {result} = renderHook<CroctProviderProps, Plug>(() => useCroct(), {
```

File: src/hooks/useCroct.test.tsx [hunk 9]
```diff
@@ -16,8 +21,10 @@ describe('useCroct', () => {
     });
 
     it('should return the Plug instance', () => {
-        const {result} = renderHook(() => useCroct(), {
-            wrapper: ({children}) => (<CroctContext.Provider value={{plug: croct}}>{children}</CroctContext.Provider>),
+        const {result} = renderHook<RenderHookProps, Plug>(() => useCroct(), {
```
Comment (georgekaran) [inline src/hooks/useCroct.test.tsx:24]: Same here

**Review by georgekaran — CHANGES_REQUESTED** [review]

File: src/hooks/useCroct.ssr.test.tsx (hunk 8 above)
Comment (tiagobarros01) [reply src/hooks/useCroct.ssr.test.tsx]: Nice!

**Review by fboechats — APPROVED** [review]

**Review by marcospassos — APPROVED** [review]

**Comment by joaopaulomoraes** [comment]
Thanks, @marcospassos, @tiagobarros01, @georgekaran, @fboechats 💚

**Pull request merged**